File: src/services/farmacia.service.ts
/**
 * Servicio de farmacias.
 * Búsqueda, detalle y listado por geolocalización.
 */
import apiClient from './api-client';
import { Farmacia, FarmaciaDetalle } from '../types';

export interface BusquedaParams {
  external_id?: string;
  nombre?: string;
  direccion?: string;
  poblacion?: string;
  provincia?: string;
  lat?: number;
  lng?: number;
  radio_km?: number;
}

export const farmaciaService = {
  /**
   * Buscar farmacias con filtros y/o geolocalización.
   */
  async buscar(params: BusquedaParams): Promise<Farmacia[]> {
    const response = await apiClient.get<{ data: Farmacia[] }>('/farmacias', { params });
    return response.data.data;
  },

  /**
   * Obtener detalle de una farmacia (incluye post-it y estado cuestionario).
   */
  async getDetalle(farmaciaId: number): Promise<FarmaciaDetalle> {
    const response = await apiClient.get<{ data: FarmaciaDetalle }>(`/farmacias/${farmaciaId}`);
    return response.data.data;
  },
};

File: lib/services/farmacia_service.dart
import '../core/api_client.dart';
import '../models/farmacia.dart';

/// Parámetros de búsqueda de farmacias.
class BusquedaParams {
  final String? externalId;
  final String? nombre;
  final String? direccion;
  final String? poblacion;
  final String? provincia;

  const BusquedaParams({
    this.externalId,
    this.nombre,
    this.direccion,
    this.poblacion,
    this.provincia,
  });

  Map<String, dynamic> toQueryParams() {
    final params = <String, dynamic>{};
    if (externalId?.isNotEmpty == true) params['external_id'] = externalId;
    if (nombre?.isNotEmpty == true) params['nombre'] = nombre;
    if (direccion?.isNotEmpty == true) params['direccion'] = direccion;
    if (poblacion?.isNotEmpty == true) params['poblacion'] = poblacion;
    if (provincia?.isNotEmpty == true) params['provincia'] = provincia;
    return params;
  }

  bool get isEmpty => toQueryParams().isEmpty;
}

/// Servicio de farmacias (peticiones_visibilidad).
class FarmaciaService {
  /// Buscar farmacias con filtros de texto.
  /// Solo devuelve las asignadas al usuario autenticado.
  static Future<List<Farmacia>> buscar(BusquedaParams params) async {
    final response = await apiClient.get(
      '/farmacias/list',
      queryParameters: params.toQueryParams(),
    );
    final data = response.data['data'] as List<dynamic>;
    return data.map((e) => Farmacia.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Obtener detalle de una farmacia.
  static Future<FarmaciaDetalle> getDetalle(int peticionId) async {
    final response = await apiClient.get(
      '/farmacias/detail',
      queryParameters: {'id': peticionId},
    );
    return FarmaciaDetalle.fromJson(
      response.data['data'] as Map<String, dynamic>,
    );
  }
}
